aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -291,6 +292,52 @@ Tensor& sum_out(Tensor& result, const Tensor& self, DimnameList dim,
   return at::sum_out(result, self, dimnames_to_positions(self, dim), keepdim, opt_dtype);
 }
 
+Tensor& nansum_out(Tensor& result, const Tensor& self, IntArrayRef dim,
+                       bool keepdim, optional<ScalarType> opt_dtype) {
+
+  auto get_computation_dtype = [](const Tensor& self) {
+    ScalarType src_type = self.scalar_type();
+    if (at::isIntegralType(src_type, /*includeBool=*/true)) {
+      return c10::typeMetaToScalarType(c10::get_default_dtype());
+    }
+    return src_type;
+  };
+
+  ScalarType scalarType =
+      opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  // numpy promotes all integral types to `Long` and preserver the floating types.
+  const auto result_type = at::isIntegralType(scalarType) ? ScalarType::Long : scalarType;
+  // use default float type for computation if `self.dtype()` is integral
```

Comment:
Why?

Reply:
Thanks! Great point.

 I was shortsighted in deciding to only use the `nansum` stub and had to hack this logic.

Now I have updated the code to dispatch to the `sum` implementation if the dtype of input tensor is integral (as `nan` is only supported by floating points) . For floats we take the `nansum` kernel path.